programs/zk-token-proof/src/lib.rs

```diff
@@ -40,24 +42,59 @@ where
     let transaction_context = &invoke_context.transaction_context;
     let instruction_context = transaction_context.get_current_instruction_context()?;
     let instruction_data = instruction_context.get_instruction_data();
-    let proof_data = ProofInstruction::proof_data::<T, U>(instruction_data).ok_or_else(|| {
-        ic_msg!(invoke_context, "invalid proof data");
-        InstructionError::InvalidInstructionData
-    })?;
+
+    // number of accessed accounts so far
+    let mut accessed_accounts = 0_u16;
+
+    // if instruction data is exactly 4 bytes, then read proof from an account
+    let proof_data = if instruction_data.len() == PROOF_OFFSET_LENGTH {
+        let proof_data_account = instruction_context
+            .try_borrow_instruction_account(transaction_context, accessed_accounts)?;
+        accessed_accounts = accessed_accounts.checked_add(1).unwrap();
+
+        let proof_data_offset = u32::from_le_bytes(
+            instruction_data[1..PROOF_OFFSET_LENGTH]
+                .try_into()
+                .map_err(|_| InstructionError::InvalidInstructionData)?,
+        );
+        let proof_data_start: usize = proof_data_offset
+            .try_into()
+            .map_err(|_| InstructionError::InvalidInstructionData)?;
+        let proof_data_end = proof_data_start
+            .checked_add(std::mem::size_of::<T>())
+            .ok_or(InstructionError::InvalidInstructionData)?;
+        let proof_data = proof_data_account
+            .get_data()
+            .get(proof_data_start..proof_data_end)
+            .ok_or(InstructionError::InvalidAccountData)?;
+
+        *bytemuck::try_from_bytes::<T>(proof_data).map_err(|_| {
+            ic_msg!(invoke_context, "invalid proof data");
+            InstructionError::InvalidInstructionData
+        })?
+    } else {
+        *ProofInstruction::proof_data::<T, U>(instruction_data).ok_or_else(|| {
+            ic_msg!(invoke_context, "invalid proof data");
+            InstructionError::InvalidInstructionData
+        })?
```

Comment:
To avoid cloning the whole proof data here, how about verifying the proof in each branch, and then have it return the context? That way you're copying a lot less data, right?

Reply:
Yep, very good idea!